core/queryalgebra/evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/QueryBindingSetTest.java

```diff
@@ -46,6 +54,136 @@ public void testEqualsMapBindingSet() {
 		assertEquals(qbs, mbs);
 	}
 
+	@Test
+	public void testAddAll() {
+		{
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addAll(new QueryBindingSet());
+			assertEquals(0, bs.size());
+		}
+		{
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			bs.addAll(new QueryBindingSet());
+			assertEquals(1, bs.size());
+		}
+		{
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			final QueryBindingSet bs2 = new QueryBindingSet();
+			bs2.addBinding("foo", vf.createIRI("urn:foo"));
+			bs.addAll(bs2);
+			assertEquals(1, bs.size());
+		}
+		{
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			final QueryBindingSet bs2 = new QueryBindingSet();
+			bs2.addBinding("foo", vf.createIRI("urn:foo"));
+			bs2.addBinding("bar", vf.createIRI("urn:bar"));
+			bs.addAll(bs2);
+			assertEquals(2, bs.size());
+		}
+	}
+
+	@Test
+	public void testAdd() {
+		try {
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			fail();
+		} catch (AssertionError e) {
```

Comment:
I don't have a good solution here. I want to test the method and the method does give this behaviour.

Reply:
My point is that the expectation that the method isn't called with the same binding twice is not an enforced requirement (even if it is stated as an expectation in the javadoc). In production (where typically you don't have assertions enabled), doing this is merely discouraged and we make no guarantees about what happens, but it doesn't necessarily result in an error. So a test to check that it throws an AssertionError doesn't really add anything useful in terms of production behavior. 

Put another way: a test shouldn't break purely because of the _absense_ of assertions. This will. 

I mean it's no really big deal, if you prefer keeping it in I have no massive objections to that either.